packages/host/app/lib/gc-identity-context.ts

```diff
@@ -53,69 +105,88 @@ export default class IdentityContextWithGarbageCollection
+      visited.add(instance);
+      if (this.gcVisit(instance[this.#api.localId], consumptionGraph, cache)) {
+        if (this.#gcCandidates.has(instance[this.#api.localId])) {
+          console.log(
+            `garbage collecting instance ${instance[this.#api.localId]} (remote id: ${instance.id}) from store`,
+          );
           // brand the instance to make it easier for debugging
           (instance as unknown as any)[
             Symbol.for('__instance_detached_from_store')
           ] = true;
-          this.delete(instance.id);
+          this.delete(instance[this.#api.localId]);
+          if (instance.id) {
+            this.delete(instance.id);
+          }
         } else {
-          this.#gcCandidates.add(instance.id);
+          this.#gcCandidates.add(instance[this.#api.localId]);
         }
       } else {
-        this.#gcCandidates.delete(instance.id);
+        this.#gcCandidates.delete(instance[this.#api.localId]);
       }
     }
   }
 
   gcVisit(
-    id: string,
-    consumptionGraph: Map<string, string[]>,
-    hasSubscribers = new Map<string, boolean>(),
-  ): boolean {
-    let cached = hasSubscribers.get(id);
+    localId: string,
+    consumptionGraph: Map<string, Set<string>>,
+    cache: Map<string, boolean>,
+  ): boolean /* true = eligible for GC, false = not eligible for GC */ {
+    let remoteId = this.#localIds.get(localId);
+    let cached = cache.get(localId);
     if (cached !== undefined) {
       return cached;
     }
 
-    let subscribers = this.subscribers.get(id);
-    if (subscribers && subscribers.resources.length > 0) {
-      hasSubscribers.set(id, true);
-      return false;
+    let subscribers = [
+      ...((remoteId
+        ? this.#remoteIdSubscribers.get(remoteId)?.resources
+        : undefined) ?? []),
+      ...(this.#localIdSubscribers.get(localId)?.resources ?? []),
+    ];
+    if (subscribers.length > 0) {
+      cache.set(localId, NOT_ELIGIBLE_FOR_GC);
+      return NOT_ELIGIBLE_FOR_GC;
     }
-    let consumers = consumptionGraph.get(id);
-    if (!consumers || consumers.length === 0) {
-      hasSubscribers.set(id, false);
-      return true;
+    let consumers = consumptionGraph.get(localId);
+    if (!consumers || consumers.size === 0) {
+      cache.set(localId, ELIGIBLE_FOR_GC);
+      return ELIGIBLE_FOR_GC;
     }
 
-    let result = consumers
-      .map((c) => this.gcVisit(c, consumptionGraph, hasSubscribers))
-      .some((result) => result);
-    hasSubscribers.set(id, result);
+    // you are eligible for GC if all your consumers are also eligible for GC
+    let result = [...consumers]
+      .map((c) => this.gcVisit(c, consumptionGraph, cache))
+      .every((result) => result);
+    cache.set(localId, result);
     return result;
   }
 
-  private makeConsumptionGraph(): Map<string, string[]> {
-    let consumptionGraph = new Map<string, string[]>();
-    for (let { card: instance } of this.#cards.values()) {
+  // this consumption graph uses local ID's
```

Comment:
Is there ever a pattern where a custom type that’s just a string is used? To avoid needing this kind of comment, so developers get a hint via typing

Reply:
we could use `type LocalId = string` to make a type alias for string that helps to document our type better